File: Leetcode/129.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int sum = 0;
    void dfs(string s, TreeNode* root) {
        if (root) {
            s += root->val + '0';
            if (root->left && root->right) {
                dfs(s, root->left);
                dfs(s, root->right);
            } else if (root->left && !root->right) {
                dfs(s, root->left);
            } else if (!root->left && root->right) {
                dfs(s, root->right);
            } else {
                sum += stoi(s);
            }
        }
    }
    int sumNumbers(TreeNode* root) {
        if (!root) return 0;
        dfs(string(), root);
        return sum;
    }
};

File: Leetcode/129/129.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
var sum int

func dfs(node *TreeNode, result []int) {
	if node != nil {
		result = append(result, node.Val)
		if node.Left != nil && node.Right != nil {
			dfs(node.Left, result)
			dfs(node.Right, result)
		} else if node.Left != nil && node.Right == nil {
			dfs(node.Left, result)
		} else if node.Left == nil && node.Right != nil {
			dfs(node.Right, result)
		} else {
			if len(result) > 0 {
				runes := make([]rune, len(result))
				for i := 0; i < len(result); i++ {
					runes[i] = rune('0' + result[i])
				}
				temp, err := strconv.Atoi(string(runes))
				if err != nil {
					log.Fatalln(err)
				}
				sum += temp
			}
		}
	}
}

func sumNumbers(root *TreeNode) int {
	sum = 0
    var temp []int
	dfs(root, temp)
	return sum
}